### PR TITLE
Allow user to configure what mockwebserver binds to

### PR DESCRIPTION
**Pull request by brennantaylor**
Provide `start(InetSocketAddress)` and `start(InetAddress, port)` to
address #1342

### CODE REVIEW

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 1]
```diff
@@ -282,13 +293,11 @@ public void start() throws IOException {
    *     tests should always use port 0 to avoid flakiness when a specific port
    *     is unavailable.
    */
-  public void start(int port) throws IOException {
+  public void start(InetSocketAddress inetSocketAddress) throws IOException {
     if (executor != null) throw new IllegalStateException("start() already called");
     executor = Executors.newCachedThreadPool(Util.threadFactory("MockWebServer", false));
-    inetAddress = InetAddress.getByName("localhost");
     serverSocket = serverSocketFactory.createServerSocket();
-    serverSocket.setReuseAddress(port != 0); // Reuse the port if the port number was specified.
```
Comment (brennantaylor) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: I think I have an idea to fix this.

**Comment by brennantaylor** [comment]
Whoops docs are out of date.

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 2]
```diff
@@ -276,22 +278,44 @@ public void start() throws IOException {
...
    * @param port the port to listen to, or 0 for any available port. Automated
    *     tests should always use port 0 to avoid flakiness when a specific port
    *     is unavailable.
    */
   public void start(int port) throws IOException {
+    start(InetAddress.getByName("localhost"), port);
+  }
+
+  /**
+   * Starts the server on the given inetaddress and port.
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: inetaddress isn't a word. You either want "address" or to wrap it in `{@code .. }`.

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 3]
```diff
@@ -276,22 +278,44 @@ public void start() throws IOException {
...
+  public void start(InetSocketAddress inetSocketAddress) throws IOException {
     if (executor != null) throw new IllegalStateException("start() already called");
     executor = Executors.newCachedThreadPool(Util.threadFactory("MockWebServer", false));
-    inetAddress = InetAddress.getByName("localhost");
+    this.inetSocketAddress = inetSocketAddress;
     serverSocket = serverSocketFactory.createServerSocket();
-    serverSocket.setReuseAddress(port != 0); // Reuse the port if the port number was specified.
-    serverSocket.bind(new InetSocketAddress(inetAddress, port), 50);
+    serverSocket.setReuseAddress(this.inetSocketAddress.getPort() != 0);
+    serverSocket.bind(this.inetSocketAddress, 50);
```
Comment (JakeWharton) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: No need to qualify with `this` for these two references.

**Comment by brennantaylor** [comment]
@JakeWharton Addressed.

**Comment by JakeWharton** [comment]
Our CI integration is screwed up. Please see https://travis-ci.org/square/okhttp/pull_requests

**Comment by brennantaylor** [comment]
Thanks. On it.

**Comment by brennantaylor** [comment]
@JakeWharton I corrected the issue but I can't tell if Travis is running the tests.

**Comment by JakeWharton** [comment]
Can you change something and force push? CI should be fixed. There's an empty trailing line in that one Javadoc, delete that maybe?

**Comment by brennantaylor** [comment]
@JakeWharton Done looks like it is marked pending in the github ui now.

**Comment by JakeWharton** [comment]
Sweet!

**Comment by brennantaylor** [comment]
@JakeWharton I get these failures when I run locally without my changes.

File: mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java [hunk 4]
```diff
@@ -276,22 +278,44 @@ public void start() throws IOException {
   }
 
   /**
-   * Starts the server.
+   * Starts the server on the loopback interface for the given port.
```
Comment (swankjesse) [inline mockwebserver/src/main/java/com/squareup/okhttp/mockwebserver/MockWebServer.java]: Nice.

**Comment by brennantaylor** [comment]
I just got the master branch tests passing on my local machine. Looking into the failures introduced by my changes now.

**Comment by brennantaylor** [comment]
So close. openjdk7 passing, openjdk8 having issues.

**Comment by swankjesse** [comment]
FYI, we've had some flakiness issues with JDK8 lately. If you've got JDK7 passing, and you don't think the other changes are related, they probably aren't!

**Comment by brennantaylor** [comment]
I don't really have the experience to say one way or the other.

**Comment by JakeWharton** [comment]
I don't recognize the one failure as a common flake. Re-kicking the build.

**Pull request merged**